arcon/src/stream/channel/strategy/keyed.rs

```diff
@@ -92,11 +90,12 @@ where
     #[inline]
     pub fn add(&mut self, event: ArconEvent<A>) -> Vec<(Arc<Channel<A>>, ArconMessage<A>)> {
         match &event {
-            ArconEvent::Element(element) => {
+            ArconEvent::Element(e) => {
                 // Get key placement
-                let key = element.data.get_key() % self.max_key;
+                //let key = self.extractor(&element.data);
+                let key = self.key_builder.get_key(&e.data);
                 // Calculate which key range index is responsible for this key
-                let index = (key * self.key_ranges / self.max_key) as usize;
+                let index = (key % self.key_ranges) as usize; // ??
```

Comment:
With the updated partitioning code in https://github.com/cda-group/arcon/blob/master/arcon/src/data/partition.rs , this is where we should do the shard lookup after extracting the key. 

As we already have extracted the key, I reckon we might have to add another function or refactor ``shard_lookup``.

```rust
let key = self.key_builder.get_key(&e.data);
let shard/index = shard_lookup_with_key(key, self.total_shards);
```
Perhaps something like this.

```rust
// src/data/partition.rs
pub fn shard_lookup<K>(data: &K, total_shards: u32) -> u32
where
    K: Hash + ?Sized,
{
    let mut hasher = arcon_util::key_hasher();
    data.hash(&mut hasher);

    shard_lookup_with_key(hasher.finish())
}

#[inline]
pub fn shard_lookup_with_key(hashed_key: u64, total_shards: u32) -> u32 {
  let key = hashed_key % MAX_KEY;
  key * total_shards / MAX_KEY
}
```

Reply:
Ok I used that snippet. This Shard/Key/Index/MAX_KEY/total_shards needs documenting but I can't do it, it's very confusing to me.

Reply:
Yes, that is a fair point. The new shard terminology and its functions have not yet been integrated with the runtime. 
So there is a bit of mix of terms used (e.g., shard vs. index)

And yes, it does indeed need more explaination text. Can leave that for other issue.

In short.

MAX_KEY represents the max key 0...65363

first calculation is to find where the hashed key maps to in the key space (0..MAX_KEY). Then the next step
is to identify the shard responsible for that key.

But yeah let's leave the integration and documentation for other issue/pr. But could you remove the PartitionIndex type? Leaving it as u32/u64 instead.